observability/src/main/java/io/grpc/observability/interceptors/LogHelper.java

```diff
@@ -0,0 +1,375 @@
+import com.google.protobuf.ByteString;
+import com.google.protobuf.Duration;
+import com.google.protobuf.util.Timestamps;
+import io.grpc.Attributes;
+import io.grpc.Deadline;
+import io.grpc.Grpc;
+import io.grpc.InternalMetadata;
+import io.grpc.Metadata;
+import io.grpc.Status;
+import io.grpc.internal.TimeProvider;
+import io.grpc.observability.logging.Sink;
+import io.grpc.observabilitylog.v1.GrpcLogRecord;
+import io.grpc.observabilitylog.v1.GrpcLogRecord.Address;
+import io.grpc.observabilitylog.v1.GrpcLogRecord.EventType;
+import io.grpc.observabilitylog.v1.GrpcLogRecord.LogLevel;
+import java.net.Inet4Address;
+import java.net.Inet6Address;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+
+/**
+ * Helper class for GCP observability logging.
+ */
+class LogHelper {
+  private static final Logger logger = Logger.getLogger(LogHelper.class.getName());
+
+  // TODO(dnvindhya): Define it in one places(TBD) to make it easily accessible from everywhere
+  static final Metadata.Key<byte[]> STATUS_DETAILS_KEY =
+      Metadata.Key.of(
+          "grpc-status-details-bin",
+          Metadata.BINARY_BYTE_MARSHALLER);
+
+  private final Sink sink;
+  private final TimeProvider timeProvider;
+
+  LogHelper(Sink sink, TimeProvider timeProvider) {
+    this.sink = sink;
+    this.timeProvider = timeProvider;
+  }
+
+  /**
+   * Logs the request header.
+   * Binary logging equivalent of logClientHeader.
+   */
+  void logRequestHeader(
+      long seqId,
+      String serviceName,
+      String methodName,
+      String authority,
+      @Nullable Duration timeout,
+      Metadata metadata,
+      GrpcLogRecord.EventLogger eventLogger,
+      String rpcId,
+      // null on client side
+      @Nullable SocketAddress peerAddress) {
+    checkNotNull(serviceName, "serviceName");
+    checkNotNull(methodName, "methodName");
+    checkNotNull(rpcId, "rpcId");
+    checkArgument(
+        peerAddress == null || eventLogger == GrpcLogRecord.EventLogger.LOGGER_SERVER,
+        "peerAddress can only be specified by server");
+
+    PayloadBuilder<GrpcLogRecord.Metadata.Builder> pair = createMetadataProto(metadata);
+    GrpcLogRecord.Builder logEntryBuilder = createTimestamp()
+        .setSequenceId(seqId)
+        .setServiceName(serviceName)
+        .setMethodName(methodName)
+        .setAuthority(authority)
+        .setEventType(EventType.GRPC_CALL_REQUEST_HEADER)
+        .setEventLogger(eventLogger)
+        .setLogLevel(LogLevel.LOG_LEVEL_DEBUG)
+        .setMetadata(pair.proto)
+        .setPayloadSize(pair.size)
+        .setRpcId(rpcId);
+    if (timeout != null) {
+      logEntryBuilder.setTimeout(timeout);
+    }
+    if (peerAddress != null) {
+      logEntryBuilder.setPeerAddress(socketAddressToProto(peerAddress));
+    }
+    logger.log(Level.FINEST, "Writing Request Header to GcpLogSink");
```

Comment:
This really should be moved to inside `GcpLogSink.write` . There you can extract the event_type from the log-entry and print out a JUL message very similar to what you have here. And it eliminates such extra calls here

Reply:
Done.